### DISCUSSION TITLE
Direct backend class or a decorator usage without calling `setup`

### DISCUSSION THREAD

**Discussion by simon-liebehenschel** [Krukov/cashews#106] [Q&A] [answered]
Based on https://github.com/Krukov/cashews/issues/105#issuecomment-1319259021

1. @Krukov What do you think about allowing to do not call `setup` if a user inherits from a backend class?

```python
# user's file
from cashews.backends.redis import Redis as CashewsRedis

class CustomRedis(CashewsRedis): 
    ...
```
2. I hate typing strings like `"mem://?size=500"` because:
   1. there are no type hints
   2. easy to make a typo
   It may be great to allow passing args and kwargs:
   ```python
   # `backend` is `Enum` type, `size` is integer type
   cache(backend=backends.MEMORY, size=500)
   ```
   This is similar to how pymongo allows to use either a string, or args/kwargs: https://pymongo.readthedocs.io/en/stable/tutorial.html#making-a-connection-with-mongoclient
   What is your opinion?

**Comment by Krukov** [comment, marked as answer]
Hi @AIGeneratedUsername .
1. It is possible now. User don't need to call `setup` if he want to use backends directly
```python
import asyncio
import random

from cashews.backends.redis import Redis
from cashews.decorators.cache.simple import cache


async def func():
    backend = Redis("redis://localhost")
    await backend.init()

    @cache(backend=backend, ttl="10m", prefix="hash")
    async def get_token():
        ...

    print(await get_token())


asyncio.run(func())

```
`setup` is a part of the cache wrapper that handle: multibackend, middleware, autoinit, disable/enable functionality.

By the way now user can use his own backend fully integrated in cashews  by register_backend function 
```python

from cashews.backends.redis import Redis
from cashews import cache, register_backend


class MyRedis(Redis):
    pass


register_backend("mybackend", MyRedis)
cache.setup("mybackend://", address="redis://localhost")

```

2.  I know about lack of typing and may be it is not clear for users, but they may pass any parameter for backend with kwargs (not by string). Yes it is not support typing and possible parameters for backends are not clear at all. This problem is critical for redis backend because if user pass unknown parameter to setup cache will not work and may fail if safe=False.
I hope typing for parameters will be soon.

Regarding enum for backend types: Actially the same solution aiocache have but I don't like it. Most of modern production applications  follow [12factors rules ](https://12factor.net/) and use environment variables to configure it. It became ugly with enums and also If user want to set backend parameters by env variables:

```python
from os import env 
from cashews import cache

backend = {"mem": backends.MEMORY, "redis": backends.REDIS}[env.get("CACHE_BACKEND", "mem")]

cache.setup(backend, max_connections=int(env.get("REDIS_CACHE_MAX_CONNECTIONS"), client_name=env.get("REDIS_APPLICATION_NAME")

```
or even ugly ... 

That is why cashews try to solve it for user and force to use uri for configuration:

```python
from os import env 
from cashews import cache

# CACHE_URI = "redis://redis.host:4888/1?max_connections=100&client_name=application"
cache.setup(env.get("CACHE_URI"))
```

Also there is a problem with enum if I want to set my own backend 
```
register_backend("mybackend", MyRedis)
```
